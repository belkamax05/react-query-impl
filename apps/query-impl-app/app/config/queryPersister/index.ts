import { createSyncStoragePersister } from '@tanstack/query-sync-storage-persister';
import { PersistedClient } from '@tanstack/react-query-persist-client';
import { IS_CLIENT } from '../env';
import persistQueryKeysMap from '../persistQueryKeysMap';

export const persisterSerialize = (client: PersistedClient) => {
  const resultClient: PersistedClient = {
    ...client,
    clientState: {
      ...client.clientState,
      queries: client.clientState.queries.filter((query) => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const res = (persistQueryKeysMap as any)[query.queryKey.join('.') as any];
        return res;
      }),
    },
  };
  return JSON.stringify(resultClient);
};

export const persisterDeserialize = (cachedString: string) => {
  const client: PersistedClient = JSON.parse(cachedString);
  return client;
};

const queryPersister = createSyncStoragePersister({
  key: 'persist-query',
  storage: IS_CLIENT ? window.localStorage : null,
  serialize: persisterSerialize,
  deserialize: persisterDeserialize,
});

export default queryPersister;
