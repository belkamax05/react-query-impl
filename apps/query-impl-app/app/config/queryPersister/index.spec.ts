import localStorageMock from '@query-impl/core/testing/mocks/localStorageMock';
import { DehydratedState } from '@tanstack/react-query';
import { PersistedClient } from '@tanstack/react-query-persist-client';
import queryPersister, { persisterDeserialize, persisterSerialize } from '.';

Object.defineProperty(window, 'localStorage', { value: localStorageMock });

describe('queryPersister', () => {
  it('should create query persister and call createSyncStoragePersister', () => {
    expect(queryPersister).toBeDefined();
    expect(queryPersister).toMatchSnapshot();
  });
  it('should serialize data', () => {
    const result = persisterSerialize({
      clientState: {
        queries: [
          // { queryKey: ['test-key'], queryHash: 'test-hash', state: null },
        ],
      } as Partial<DehydratedState> as DehydratedState,
    } as Partial<PersistedClient> as PersistedClient);
    expect(result).toStrictEqual('{"clientState":{"queries":[]}}');
  });
  it('should deserialize data', () => {
    const result = persisterDeserialize('{"some":"data"}');
    expect(result).toStrictEqual({ some: 'data' });
  });
});
