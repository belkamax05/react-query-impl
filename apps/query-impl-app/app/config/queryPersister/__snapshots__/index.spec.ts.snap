// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryPersister should create query persister and call createSyncStoragePersister 1`] = `
{
  "persistClient": [Function],
  "removeClient": [Function],
  "restoreClient": [Function],
}
`;
